src/Lecture2.hs

```diff
@@ -163,8 +194,90 @@ data Knight = Knight
     , knightAttack    :: Int
     , knightEndurance :: Int
     }
+    deriving (Show)
+
+data Dragon = Dragon
+    { dragonHealth     :: Int
+    , dragonExpReward :: Int
+    , dragonFirePower  :: Int
+    }
+    deriving (Show)
+
+--type GreenDragon = Dragon
+--type RedDragon   = Dragon
+--type BlackDragon = Dragon
+
+data Chest a = Chest
+    { chestGold     :: Int
+    , chestTreasure :: a
+    }
+    deriving (Show)
 
-dragonFight = error "TODO"
+data Armor = Armo
+    { armorDefence :: Int
+    , armorCost    :: Int
+    }
+    deriving (Show)
+
+data Sword = Sword
+    { swordImpact :: Int
+    , swordCost   :: Int
+    }
+    deriving (Show)
+
+data Artifact = Artifact
+    { artifactBoost :: Int
+    , artifactCost  :: Int
+    }
+    deriving (Show)
+
+data Gemstone = Gemstone
+    {
+      gemstoneCost :: Int
+    }
+    deriving (Show)
+
+type BronzeChest = Chest Armor
+type SilverChest = Chest (Sword, Armor)
+type GoldenChest = Chest (Artifact, Sword, [Gemstone])
+
+
+data RewardChest
+  = Bronze BronzeChest
+  | Silver SilverChest
+  | Golden GoldenChest
+  deriving (Show)
+
+data FightResult
+  = KnightWin
```

Comment:
You can put `Chest` and `Experience` inside the `KnightWin` constructor to keep this information for future use in case of win 🙂 

Otherwise, your `RewardChest` data types are not used in the `dragonFight` function.